experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,253 @@
+                    }
+                    array[i].write(array[j].assume_init_ref().clone());
+                },
+            }
+        }
+    }
+
+    // https://github.com/rust-lang/rust/issues/61956
+    Ok(unsafe { core::ptr::read(array.as_ptr() as *const [T; N]) })
+}
+
+#[derive(serde::Serialize)]
+#[serde(untagged)]
+enum HumanSer<'a, T> {
+    Value(&'a T),
+    Fallback([usize; 1]),
+}
+
+enum MachineSer<'a, T> {
+    Value(&'a T),
+    Fallback(usize),
+}
+
+// These enums are public because they are exposed in the trait bounds of
+// dedupe_array_de. Serde only implements Deserialize<'de> on arrays up to
+// size 32, so [Dedupe<T>;N]: Deserialize<'de> cannot be inferred from
+// Dedupe<T>: Deserialize<'de> in the general case. See
+// https://github.com/serde-rs/serde/issues/1937.
+//
+// These are not considered part of the stable API.
+
+#[derive(serde::Deserialize)]
+#[serde(untagged)]
+#[doc(hidden)]
+pub enum HumanDe<T> {
+    Value(T),
+    Fallback([usize; 1]),
+}
+
+#[doc(hidden)]
+pub enum MachineDe<T> {
+    Value(T),
+    Fallback(usize),
+}
+
+impl<'a, T> Serialize for MachineSer<'a, T>
+where
+    T: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        T: Serialize,
+    {
+        match self {
+            // Serialize values as an enum variant with index 0
+            MachineSer::Value(t) => serializer.serialize_newtype_variant(
+                "unused-enum-name",
+                0,
+                "unused-enum-variant",
+                t,
+            ),
+            // Serialize fallbacks as an (empty) variant with index = fallback + 1
+            MachineSer::Fallback(fallback) => serializer.serialize_unit_variant(
+                "unused-enum-name",
+                (fallback + 1) as u32,
+                "unused-enum-variant",
+            ),
+        }
+    }
+}
+
+impl<'de, T: Deserialize<'de>> Deserialize<'de> for MachineDe<T> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct DedupeVisitor<R>(core::marker::PhantomData<R>);
+
+        impl<'de, R: Deserialize<'de>> Visitor<'de> for DedupeVisitor<R> {
+            type Value = MachineDe<R>;
+
+            fn expecting(
+                &self,
+                formatter: &mut alloc::fmt::Formatter,
+            ) -> Result<(), alloc::fmt::Error> {
+                formatter.write_str("Element or fallback reference.")
+            }
+
+            fn visit_enum<A>(self, data: A) -> Result<Self::Value, A::Error>
+            where
+                A: EnumAccess<'de>,
+            {
+                use serde::de::VariantAccess;
+
+                let (variant, variant_access) = data.variant()?;
+                Ok(match variant {
+                    0 => MachineDe::Value(variant_access.newtype_variant()?),
+                    n => MachineDe::Fallback(n - 1),
+                })
```

Comment:
Note: If you accept my suggestion above, I *think* this code would need to change to expect a `u32` as the variant for machine-readable, and a `String` as the variant for human-readable.